Sources/ApodiniObserve/ConfiguredLogger.swift

```diff
@@ -0,0 +1,232 @@
+            // Write request metadata
+            builtLogger?[metadataKey: "request"] = .dictionary(self.getRequestMetadata(from: request)
+                                                                .merging(self.getRawRequestMetadata(from: connection.information)) { _, new in new })
+            
+            // Write endpoint metadata
+            builtLogger?[metadataKey: "endpoint"] = .dictionary(self.endpointMetadata)
+            
+            // Write exporter metadata
+            builtLogger?[metadataKey: "exporter"] = .dictionary(self.exporterMetadata)
+            
+            /// Prio 1: User specifies a `Logger.LogLevel` in the property wrapper for a specific `Handler`
+            if let logLevel = self.logLevel {
+                builtLogger?.logLevel = logLevel
+                
+                // If logging level is configured gloally
+                if let globalConfiguredLogLevel = storage.get(LoggerConfiguration.LoggingStorageKey.self)?.configuration.logLevel {
+                    if logLevel < globalConfiguredLogLevel {
+                        logger.warning("The global configured logging level is \(globalConfiguredLogLevel.rawValue) but Handler \(self.blackboardMetadata.endpointName) has logging level \(logLevel.rawValue) which is lower than the configured global logging level")
+                    }
+                // If logging level is automatically set to a default value
+                } else {
+                    var globalLogLevel: Logger.Level
+                    #if DEBUG
+                    globalLogLevel = .debug
+                    #else
+                    globalLogLevel = .info
+                    #endif
+                    
+                    if logLevel < globalLogLevel {
+                        logger.warning("The global default logging level is \(globalLogLevel.rawValue) but Handler \(self.blackboardMetadata.endpointName) has logging level \(logLevel.rawValue) which is lower than the global default logging level")
+                    }
+                }
+            }
+            /// Prio 2: User specifies a `Logger.LogLevel`either via a CLI argument or via a `LoggerConfiguration` in the configuration of the `WebService`
+            else if let loggingConfiguraiton = storage.get(LoggerConfiguration.LoggingStorageKey.self)?.configuration {
+                builtLogger?.logLevel = loggingConfiguraiton.logLevel
+            }
+            /// Prio 3: No `Logger.LogLevel` specified by user, use defaul value according to environment (debug mode or release mode)
+            else {
+                #if DEBUG
+                builtLogger?.logLevel = .debug
+                #else
+                builtLogger?.logLevel = .info
+                #endif
+            }
+        } else {
+            // Not pretty, but otherwise ApodiniObserve would need to depend on ApodiniWebsocket
+            if String(describing: exporterTypeMetadata.exporterType) == "WebSocketInterfaceExporter" {
```

Comment:
I absolutely agree with you, that was just meant to be a temporary solution.

I could use something like `#if canImport(ApodiniWebSocket)`, but the `WebSocketInterfaceExporter` type isn't public anymore, therefore I would have to use the public `WebSocket`struct and attach `InterfaceExporter` to it, also not very pretty, but at least a little better. 
I also tried to just save the type of the `Exporter` to the the `WebSocket.ExporterConfiguration` (as a static property), but somehow this results in quite weird and ever-changing errors.

``
#if canImport(ApodiniWebSocket)
            if String(describing: exporterTypeMetadata.exporterType) == String(describing: ApodiniWebSocket.WebSocket.ExporterConfiguration.exporterType) {
                // Write connection state
                builtLogger?[metadataKey: "connectionState"] = .string(connection.state.rawValue)
                
                // Write request metadata
                builtLogger?[metadataKey: "request"] = .dictionary(self.getRequestMetadata(from: connection.request))
            }
            #endif
``
Do you have any other suggestions for me on how to easily achieve smt. like that? :)

Reply:
Hm I see. I'm okay with the current approach if it stays temporary.

What I had in mind was to think the other way round. Providing a generalized way of doing such "updates" to the logger. I could see two possible approaches:
* ApodiniObserve provides some sort of hooks which are executed on such occasions. This could be realized with some sort of global storage maybe. The WebSocket Exporter would register such hooks with ApodiniObserve and those would then be executed on the right place to update the Loggers metadata. Thinking that thought to the end, it might not be the best solution to this exact problem, as it might actually be a problem for the WebSocket exporter to compute the updated values.
* Thinking about it. We really only need to update the `connectionState` and `request` metadataKeys. Why? Because the WebSocket exporter uses a `bidirectionalStream` CommunicationalPattern when exporting its endpoints. And its not only the WebSocket Exporter, I assume the update of those logger metadata entires would need to be also done for the plain HTTP exporter if it uses the `bidirectionalStream` CommunicationalPattern (The same thing probably also applies to the `clientSideStream` pattern 🤔). @theMomax can we - specifically the `ConfiguredLogger` property - detect which CommunicationalPattern was used by the Exporter to export the currently executed Handler? We have the `CommunicationalPattern` Metadata, but this only represents a request to the exporter which pattern should be used and doesn't necessarily result in that pattern being exported, right?  
Thinking about it, we really only want to update those two keys if they changed right? For the `connectionState` it is probably relative simple, as overwriting a single value on each logger access isn't probably that huge of an efficiency problem. But can we efficiently detect if the request changed, to not compute the whole metadata value on every logger access?

Reply:
Also, i think there's not really an efficient way to detect a change to the request (eg. we're aggregating some logging metadata data from Vapor, I don't think there's even a possibility to detect changes without evaluation then, right?)

Reply:
Ah yeah, looking into the `ProtocolAgnosticEndpointKnowledge`more closely, I came to that conclusion too. Thanks Max!